### PR TITLE
Don't make new IR nodes if nothing changed when simplifying

### PR DESCRIPTION
**Pull request by abadams**
Some statements and expressions get crafted anew each time when
repeatedly resimplified. This PR fixes all the cases I could find.

No significant changes to most pipelines. Reduces total compile time by 7% on one particularly pathological example.

### CODE REVIEW

**Review by steven-johnson — APPROVED** [review]

File: src/Simplify_EQ.cpp [hunk 1]
```diff
@@ -106,25 +106,37 @@ Expr Simplify::visit(const EQ *op, ExprInfo *bounds) {
...
+        return op;
+    } else if (add &&
+               add->a.same_as(op->a) &&
+               (op->b.as<IntImm>() ||
+                op->b.as<UIntImm>() ||
+                op->b.as<FloatImm>())) {
+        // a - b became a + c. c must be the negation of b, which
+        // can't possibly get simpler.
+        return op;
+    } else if (delta.same_as(op->a) &&
```
Comment (rootjalex) [inline src/Simplify_EQ.cpp:126]: Having this check before the `rewrite(x - y == 0, x == y)` below means that `a - b == 0` (assuming `a - b` doesn't simplify further) would be returned as simply `a - b == 0`, and never simplified to `a == b`

Comment (abadams) [reply src/Simplify_EQ.cpp]: Good point. That must be why the tests are failing. I pushed a simpler alternative (though it does construct a another temporary Expr, sadly)

**Comment by steven-johnson** [comment]
LGTM

File: src/Simplify_EQ.cpp (hunk 1 above)
Comment (rootjalex) [reply src/Simplify_EQ.cpp]: I think the first two conditions could actually be left in, which would still refrain from constructing temporary Exprs in those cases at least? I think just the last one causes problems.

**Pull request merged**